File: ios/WiretalkSDK/Sources/WiretalkSDK/WiretalkConfig.swift
import Foundation

public enum WiretalkPlatform: String {
    case android
    case ios
    case reactNative = "react-native"
    case flutter
    case capacitor
    case cordova
}

public struct WiretalkConfig {
    public let baseURL: URL
    public let widgetKey: String
    public let appId: String
    public let platform: WiretalkPlatform
    public let locale: String
    public let initialTab: String
    public let visitorUid: String?
    public let visitorName: String?
    public let visitorEmail: String?
    public let visitorPhone: String?
    public let appVersion: String?
    public let appBuild: String?

    public init(
        baseURL: URL,
        widgetKey: String,
        appId: String,
        platform: WiretalkPlatform = .ios,
        locale: String = "en",
        initialTab: String = "chat",
        visitorUid: String? = nil,
        visitorName: String? = nil,
        visitorEmail: String? = nil,
        visitorPhone: String? = nil,
        appVersion: String? = nil,
        appBuild: String? = nil
    ) {
        self.baseURL = baseURL
        self.widgetKey = widgetKey
        self.appId = appId
        self.platform = platform
        self.locale = locale
        self.initialTab = initialTab
        self.visitorUid = visitorUid
        self.visitorName = visitorName
        self.visitorEmail = visitorEmail
        self.visitorPhone = visitorPhone
        self.appVersion = appVersion
        self.appBuild = appBuild
    }

    public func popoutURL() -> URL {
        var components = URLComponents(url: baseURL.appendingPathComponent("popout/\(widgetKey)"), resolvingAgainstBaseURL: false)!
        var items = [
            URLQueryItem(name: "locale", value: locale),
            URLQueryItem(name: "tab", value: initialTab),
            URLQueryItem(name: "platform", value: platform.rawValue),
            URLQueryItem(name: "app_id", value: appId),
        ]
        if let visitorUid, !visitorUid.isEmpty { items.append(URLQueryItem(name: "visitor_uid", value: visitorUid)) }
        if let visitorName, !visitorName.isEmpty { items.append(URLQueryItem(name: "visitor_name", value: visitorName)) }
        if let visitorEmail, !visitorEmail.isEmpty { items.append(URLQueryItem(name: "visitor_email", value: visitorEmail)) }
        if let visitorPhone, !visitorPhone.isEmpty { items.append(URLQueryItem(name: "visitor_phone", value: visitorPhone)) }
        if let appVersion, !appVersion.isEmpty { items.append(URLQueryItem(name: "app_version", value: appVersion)) }
        if let appBuild, !appBuild.isEmpty { items.append(URLQueryItem(name: "app_build", value: appBuild)) }
        components.queryItems = items
        return components.url!
    }

    public static func generateVisitorUid() -> String {
        UUID().uuidString.lowercased()
    }
}

File: android/wiretalk-android/src/main/java/com/wiretalk/sdk/WiretalkConfig.kt
package com.wiretalk.sdk

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Parcelable
import kotlinx.parcelize.Parcelize
import java.util.UUID

enum class WiretalkPlatform(val queryValue: String) {
    ANDROID("android"),
    IOS("ios"),
    REACT_NATIVE("react-native"),
    FLUTTER("flutter"),
    CAPACITOR("capacitor"),
    CORDOVA("cordova"),
}

@Parcelize
data class WiretalkConfig(
    val baseUrl: String,
    val widgetKey: String,
    val appId: String,
    val platform: WiretalkPlatform = WiretalkPlatform.ANDROID,
    val locale: String = "en",
    val initialTab: String = "chat",
    val visitorUid: String? = null,
    val visitorName: String? = null,
    val visitorEmail: String? = null,
    val visitorPhone: String? = null,
    val appVersion: String? = null,
    val appBuild: String? = null,
) : Parcelable {
    fun popoutUrl(): String {
        val base = baseUrl.trimEnd('/')
        val params = linkedMapOf(
            "locale" to locale,
            "tab" to initialTab,
            "platform" to platform.queryValue,
            "app_id" to appId,
        )
        visitorUid?.takeIf { it.isNotBlank() }?.let { params["visitor_uid"] = it }
        visitorName?.takeIf { it.isNotBlank() }?.let { params["visitor_name"] = it }
        visitorEmail?.takeIf { it.isNotBlank() }?.let { params["visitor_email"] = it }
        visitorPhone?.takeIf { it.isNotBlank() }?.let { params["visitor_phone"] = it }
        appVersion?.takeIf { it.isNotBlank() }?.let { params["app_version"] = it }
        appBuild?.takeIf { it.isNotBlank() }?.let { params["app_build"] = it }

        val query = params.entries.joinToString("&") { (key, value) ->
            "${Uri.encode(key)}=${Uri.encode(value)}"
        }

        return "$base/popout/${Uri.encode(widgetKey)}?$query"
    }

    fun withVisitorUid(uid: String?) = copy(visitorUid = uid?.takeIf { it.isNotBlank() })

    companion object {
        fun generateVisitorUid(): String = UUID.randomUUID().toString()
    }
}

object WiretalkChatActivity {
    private const val EXTRA_CONFIG = "wiretalk_config"

    fun launch(context: Context, config: WiretalkConfig) {
        val intent = Intent(context, WiretalkChatActivityImpl::class.java).apply {
            putExtra(EXTRA_CONFIG, config)
        }
        context.startActivity(intent)
    }

    internal fun readConfig(intent: Intent?): WiretalkConfig? {
        return intent?.getParcelableExtra(EXTRA_CONFIG)
    }
}
